check_nodes.go

```diff
@@ -142,16 +147,37 @@ func checkNodes(ctx context.Context, nodes []Node, executor checkExecutorFunc, t
 	}
 }
 
-// checkExecutor returns checkExecutorFunc which can execute supplied check.
-func checkExecutor(checker NodeChecker) checkExecutorFunc {
-	return func(ctx context.Context, node Node) (bool, time.Duration, error) {
+// checkRoleExecutor returns checkExecutorFunc which can execute role check.
+func checkRoleExecutor(checker NodeChecker) checkExecutorFunc {
+	return func(ctx context.Context, target *checkedNode) error {
 		ts := time.Now()
-		primary, err := checker(ctx, node.DB())
+		primary, err := checker(ctx, target.Node.DB())
 		d := time.Since(ts)
 		if err != nil {
-			return false, d, err
+			return fmt.Errorf("unable to check node role: %w", err)
+		}
+
+		target.Primary = primary
+		target.Latency = d
+
+		return nil
+	}
+}
+
+// checkReplicationLagExecutor returns checkExecutorFunc which can execute replication lag check.
+func checkReplicationLagExecutor(checker ReplicationLagChecker, maxLag time.Duration) checkExecutorFunc {
+	return func(ctx context.Context, target *checkedNode) error {
+		if checker == nil || maxLag == 0 || target.Primary {
+			return nil
 		}
 
-		return primary, d, nil
+		lag, err := checker(ctx, target.Node.DB())
+		if err != nil {
+			return fmt.Errorf("cannot check node replication lag: %w", err)
+		}
+		if lag < maxLag {
+			return fmt.Errorf("replication lag is too big: %s", lag)
```

Comment:
We need typed errors since tracer.NodeDead can return error for both 'cant check' and 'too much lag'.

The question is do we need to create an actualy type for replication lag? User might want to know the exact lag. Or he might not. Sentinels are a bit easier to work with but provide less info.